JSON & YAML output for kubectl api-resources

dharmit opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->
/kind feature

#### What this PR does / why we need it:
JSON & YAML output for kubectl api-resources

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes https://github.com/kubernetes/kubectl/issues/1382

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Adds machine readable output options (JSON & YAML) to kubectl api-resources
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

dharmit commented:
<details>

<summary> Unit tests fail like below: </summary>

```
$ make test WHAT=./staging/src/k8s.io/kubectl/pkg/cmd/apiresources
+++ [0724 21:02:23] Set GOMAXPROCS automatically to 24
+++ [0724 21:02:23] Running tests without code coverage and with -race
--- FAIL: TestAPIResourcesRun (0.02s)
    --- FAIL: TestAPIResourcesRun/no_headers (0.00s)
        apiresources_test.go:389: unexpected output: bars            v1             true    Bar
            bars            v1             true    Bar
            foos     f,fo   v1             false   Foo
            foos     f,fo   v1             false   Foo
            bazzes   b      somegroup/v1   true    Baz
            bazzes   b      somegroup/v1   true    Baz

            expected: bars            v1             true    Bar
            foos     f,fo   v1             false   Foo
            bazzes   b      somegroup/v1   true    Baz
    --- FAIL: TestAPIResourcesRun/specific_api_group (0.00s)
        apiresources_test.go:389: unexpected output: NAME     SHORTNAMES   APIVERSION     NAMESPACED   KIND
            bars                  v1             true         Bar
            bars                  v1             true         Bar
            foos     f,fo         v1             false        Foo
            foos     f,fo         v1             false        Foo
            bazzes   b            somegroup/v1   true         Baz
            bazzes   b            somegroup/v1   true         Baz
            bazzes   b            somegroup/v1   true         Baz

            expected: NAME     SHORTNAMES   APIVERSION     NAMESPACED   KIND
            bazzes   b            somegroup/v1   true         Baz
    --- FAIL: TestAPIResourcesRun/output_wide (0.00s)
        apiresources_test.go:389: unexpected output: NAME     SHORTNAMES   APIVERSION     NAMESPACED   KIND   VERBS                    CATEGORIES
            bars                  v1             true         Bar    get,list,create
            bars                  v1             true         Bar    get,list,create
            bars                  v1             true         Bar    get,list,create
            foos     f,fo         v1             false        Foo    get,list                 some-category
            foos     f,fo         v1             false        Foo    get,list                 some-category
            foos     f,fo         v1             false        Foo    get,list                 some-category
            bazzes   b            somegroup/v1   true         Baz    get,list,create,delete   some-category,another-category
            bazzes   b            somegroup/v1   true         Baz    get,list,create,delete   some-category,another-category
            bazzes   b            somegroup/v1   true         Baz    get,list,create,delete   some-category,another-category
            bazzes   b            somegroup/v1   true         Baz    get,list,create,delete   some-category,another-category

            expected: NAME     SHORTNAMES   APIVERSION     NAMESPACED   KIND   VERBS                    CATEGORIES
            bars                  v1             true         Bar    get,list,create
            foos     f,fo         v1             false        Foo    get,list                 some-category
            bazzes   b            somegroup/v1   true         Baz    get,list,create,delete   some-category,another-category
    --- FAIL: TestAPIResourcesRun/output_name (0.00s)
        apiresources_test.go:389: unexpected output: bars
            bars
            bars
            bars
            foos
            foos
            foos
            foos
            bazzes.somegroup
            bazzes.somegroup
            bazzes.somegroup
            bazzes.somegroup
            bazzes.somegroup

            expected: bars
            foos
            bazzes.somegroup
    --- FAIL: TestAPIResourcesRun/output_json (0.00s)
        apiresources_test.go:389: unexpected output: {
                "kind": "v1",
                "apiVersion": "APIResourceList",
                "groupVersion": "",
                "resources": [
                    {
                        "name": "foos",
                        "singularName": "",
                        "namespaced": false,
                        "kind": "Foo",
                        "verbs": [
                            "get",
                            "list"
                        ],
                        "shortNames": [
                            "f",
                            "fo"
                        ],
                        "categories": [
                            "some-category"
                        ]
                    },
                    {
                        "name": "bars",
                        "singularName": "",
                        "namespaced": true,
                        "kind": "Bar",
                        "verbs": [
                            "get",
                            "list",
                            "create"
                        ]
                    },
                    {
                        "name": "bazzes",
                        "singularName": "",
                        "namespaced": true,
                        "kind": "Baz",
                        "verbs": [
                            "get",
                            "list",
                            "create",
                            "delete"
                        ],
                        "shortNames": [
                            "b"
                        ],
                        "categories": [
                            "some-category",
                            "another-category"
                        ]
                    },
                    {
                        "name": "NoVerbs",
                        "singularName": "",
                        "namespaced": true,
                        "kind": "NoVerbs",
                        "verbs": [],
                        "shortNames": [
                            "b"
                        ]
                    }
                ]
            }

            expected: {
            	"kind": "v1",
            	"apiVersion": "APIResourceList",
            	"groupVersion": "",
            	"resources": [
            		{
            			"name": "foos",
            			"singularName": "",
            			"namespaced": false,
            			"kind": "Foo",
            			"verbs": [
            				"get",
            				"list"
            			],
            			"shortNames": [
            				"f",
            				"fo"
            			],
            			"categories": [
            				"some-category"
            			]
            		},
            		{
            			"name": "bars",
            			"singularName": "",
            			"namespaced": true,
            			"kind": "Bar",
            			"verbs": [
            				"get",
            				"list",
            				"create"
            			]
            		},
            		{
            			"name": "bazzes",
            			"singularName": "",
            			"namespaced": true,
            			"kind": "Baz",
            			"verbs": [
            				"get",
            				"list",
            				"create",
            				"delete"
            			],
            			"shortNames": [
            				"b"
            			],
            			"categories": [
            				"some-category",
            				"another-category"
            			]
            		},
            		{
            			"name": "NoVerbs",
            			"singularName": "",
            			"namespaced": true,
            			"kind": "NoVerbs",
            			"verbs": [],
            			"shortNames": [
            				"b"
            			]
            		}
            	]
            }
    --- FAIL: TestAPIResourcesRun/namespaced (0.00s)
        apiresources_test.go:389: unexpected output: NAME     SHORTNAMES   APIVERSION     NAMESPACED   KIND
            bars                  v1             true         Bar
            bars                  v1             true         Bar
            bars                  v1             true         Bar
            bars                  v1             true         Bar
            bars                  v1             true         Bar
            bars                  v1             true         Bar
            foos     f,fo         v1             false        Foo
            foos     f,fo         v1             false        Foo
            foos     f,fo         v1             false        Foo
            foos     f,fo         v1             false        Foo
            foos     f,fo         v1             false        Foo
            bazzes   b            somegroup/v1   true         Baz
            bazzes   b            somegroup/v1   true         Baz
            bazzes   b            somegroup/v1   true         Baz
            bazzes   b            somegroup/v1   true         Baz
            bazzes   b            somegroup/v1   true         Baz
            bazzes   b            somegroup/v1   true         Baz
            bazzes   b            somegroup/v1   true         Baz

            expected: NAME     SHORTNAMES   APIVERSION     NAMESPACED   KIND
            bars                  v1             true         Bar
            bazzes   b            somegroup/v1   true         Baz
    --- FAIL: TestAPIResourcesRun/single_verb (0.00s)
        apiresources_test.go:389: unexpected output: NAME     SHORTNAMES   APIVERSION     NAMESPACED   KIND
            bars                  v1             true         Bar
            bars                  v1             true         Bar
            bars                  v1             true         Bar
            bars                  v1             true         Bar
            bars                  v1             true         Bar
            bars                  v1             true         Bar
            bars                  v1             true         Bar
            foos     f,fo         v1             false        Foo
            foos     f,fo         v1             false        Foo
            foos     f,fo         v1             false        Foo
            foos     f,fo         v1             false        Foo
            foos     f,fo         v1             false        Foo
            bazzes   b            somegroup/v1   true         Baz
            bazzes   b            somegroup/v1   true         Baz
            bazzes   b            somegroup/v1   true         Baz
            bazzes   b            somegroup/v1   true         Baz
            bazzes   b            somegroup/v1   true         Baz
            bazzes   b            somegroup/v1   true         Baz
            bazzes   b            somegroup/v1   true         Baz
            bazzes   b            somegroup/v1   true         Baz

            expected: NAME     SHORTNAMES   APIVERSION     NAMESPACED   KIND
            bars                  v1             true         Bar
            bazzes   b            somegroup/v1   true         Baz
    --- FAIL: TestAPIResourcesRun/multiple_verbs (0.00s)
        apiresources_test.go:389: unexpected output: NAME     SHORTNAMES   APIVERSION     NAMESPACED   KIND
            bars                  v1             true         Bar
            bars                  v1             true         Bar
            bars                  v1             true         Bar
            bars                  v1             true         Bar
            bars                  v1             true         Bar
            bars                  v1             true         Bar
            bars                  v1             true         Bar
            foos     f,fo         v1             false        Foo
            foos     f,fo         v1             false        Foo
            foos     f,fo         v1             false        Foo
            foos     f,fo         v1             false        Foo
            foos     f,fo         v1             false        Foo
            bazzes   b            somegroup/v1   true         Baz
            bazzes   b            somegroup/v1   true         Baz
            bazzes   b            somegroup/v1   true         Baz
            bazzes   b            somegroup/v1   true         Baz
            bazzes   b            somegroup/v1   true         Baz
            bazzes   b            somegroup/v1   true         Baz
            bazzes   b            somegroup/v1   true         Baz
            bazzes   b            somegroup/v1   true         Baz
            bazzes   b            somegroup/v1   true         Baz

            expected: NAME     SHORTNAMES   APIVERSION     NAMESPACED   KIND
            bazzes   b            somegroup/v1   true         Baz
    --- FAIL: TestAPIResourcesRun/single_category (0.00s)
        apiresources_test.go:389: unexpected output: NAME     SHORTNAMES   APIVERSION     NAMESPACED   KIND
            bars                  v1             true         Bar
            bars                  v1             true         Bar
            bars                  v1             true         Bar
            bars                  v1             true         Bar
            bars                  v1             true         Bar
            bars                  v1             true         Bar
            bars                  v1             true         Bar
            foos     f,fo         v1             false        Foo
            foos     f,fo         v1             false        Foo
            foos     f,fo         v1             false        Foo
            foos     f,fo         v1             false        Foo
            foos     f,fo         v1             false        Foo
            foos     f,fo         v1             false        Foo
            bazzes   b            somegroup/v1   true         Baz
            bazzes   b            somegroup/v1   true         Baz
            bazzes   b            somegroup/v1   true         Baz
            bazzes   b            somegroup/v1   true         Baz
            bazzes   b            somegroup/v1   true         Baz
            bazzes   b            somegroup/v1   true         Baz
            bazzes   b            somegroup/v1   true         Baz
            bazzes   b            somegroup/v1   true         Baz
            bazzes   b            somegroup/v1   true         Baz
            bazzes   b            somegroup/v1   true         Baz

            expected: NAME     SHORTNAMES   APIVERSION     NAMESPACED   KIND
            foos     f,fo         v1             false        Foo
            bazzes   b            somegroup/v1   true         Baz
    --- FAIL: TestAPIResourcesRun/multiple_categories (0.00s)
        apiresources_test.go:389: unexpected output: NAME     SHORTNAMES   APIVERSION     NAMESPACED   KIND
            bars                  v1             true         Bar
            bars                  v1             true         Bar
            bars                  v1             true         Bar
            bars                  v1             true         Bar
            bars                  v1             true         Bar
            bars                  v1             true         Bar
            bars                  v1             true         Bar
            foos     f,fo         v1             false        Foo
            foos     f,fo         v1             false        Foo
            foos     f,fo         v1             false        Foo
            foos     f,fo         v1             false        Foo
            foos     f,fo         v1             false        Foo
            foos     f,fo         v1             false        Foo
            bazzes   b            somegroup/v1   true         Baz
            bazzes   b            somegroup/v1   true         Baz
            bazzes   b            somegroup/v1   true         Baz
            bazzes   b            somegroup/v1   true         Baz
            bazzes   b            somegroup/v1   true         Baz
            bazzes   b            somegroup/v1   true         Baz
            bazzes   b            somegroup/v1   true         Baz
            bazzes   b            somegroup/v1   true         Baz
            bazzes   b            somegroup/v1   true         Baz
            bazzes   b            somegroup/v1   true         Baz
            bazzes   b            somegroup/v1   true         Baz

            expected: NAME     SHORTNAMES   APIVERSION     NAMESPACED   KIND
            bazzes   b            somegroup/v1   true         Baz
    --- FAIL: TestAPIResourcesRun/sort_by_name (0.00s)
        apiresources_test.go:389: unexpected output: NAME     SHORTNAMES   APIVERSION     NAMESPACED   KIND
            bars                  v1             true         Bar
            bars                  v1             true         Bar
            bars                  v1             true         Bar
            bars                  v1             true         Bar
            bars                  v1             true         Bar
            bars                  v1             true         Bar
            bars                  v1             true         Bar
            bars                  v1             true         Bar
            bazzes   b            somegroup/v1   true         Baz
            bazzes   b            somegroup/v1   true         Baz
            bazzes   b            somegroup/v1   true         Baz
            bazzes   b            somegroup/v1   true         Baz
            bazzes   b            somegroup/v1   true         Baz
            bazzes   b            somegroup/v1   true         Baz
            bazzes   b            somegroup/v1   true         Baz
            bazzes   b            somegroup/v1   true         Baz
            bazzes   b            somegroup/v1   true         Baz
            bazzes   b            somegroup/v1   true         Baz
            bazzes   b            somegroup/v1   true         Baz
            bazzes   b            somegroup/v1   true         Baz
            foos     f,fo         v1             false        Foo
            foos     f,fo         v1             false        Foo
            foos     f,fo         v1             false        Foo
            foos     f,fo         v1             false        Foo
            foos     f,fo         v1             false        Foo
            foos     f,fo         v1             false        Foo
            foos     f,fo         v1             false        Foo

            expected: NAME     SHORTNAMES   APIVERSION     NAMESPACED   KIND
            bars                  v1             true         Bar
            bazzes   b            somegroup/v1   true         Baz
            foos     f,fo         v1             false        Foo
    --- FAIL: TestAPIResourcesRun/sort_by_kind (0.00s)
        apiresources_test.go:389: unexpected output: NAME     SHORTNAMES   APIVERSION     NAMESPACED   KIND
            bars                  v1             true         Bar
            bars                  v1             true         Bar
            bars                  v1             true         Bar
            bars                  v1             true         Bar
            bars                  v1             true         Bar
            bars                  v1             true         Bar
            bars                  v1             true         Bar
            bars                  v1             true         Bar
            bars                  v1             true         Bar
            bazzes   b            somegroup/v1   true         Baz
            bazzes   b            somegroup/v1   true         Baz
            bazzes   b            somegroup/v1   true         Baz
            bazzes   b            somegroup/v1   true         Baz
            bazzes   b            somegroup/v1   true         Baz
            bazzes   b            somegroup/v1   true         Baz
            bazzes   b            somegroup/v1   true         Baz
            bazzes   b            somegroup/v1   true         Baz
            bazzes   b            somegroup/v1   true         Baz
            bazzes   b            somegroup/v1   true         Baz
            bazzes   b            somegroup/v1   true         Baz
            bazzes   b            somegroup/v1   true         Baz
            bazzes   b            somegroup/v1   true         Baz
            foos     f,fo         v1             false        Foo
            foos     f,fo         v1             false        Foo
            foos     f,fo         v1             false        Foo
            foos     f,fo         v1             false        Foo
            foos     f,fo         v1             false        Foo
            foos     f,fo         v1             false        Foo
            foos     f,fo         v1             false        Foo
            foos     f,fo         v1             false        Foo

            expected: NAME     SHORTNAMES   APIVERSION     NAMESPACED   KIND
            bars                  v1             true         Bar
            bazzes   b            somegroup/v1   true         Baz
            foos     f,fo         v1             false        Foo
    --- FAIL: TestAPIResourcesRun/cached (0.00s)
        apiresources_test.go:389: unexpected output: NAME     SHORTNAMES   APIVERSION     NAMESPACED   KIND
            bars                  v1             true         Bar
            bars                  v1             true         Bar
            bars                  v1             true         Bar
            bars                  v1             true         Bar
            bars                  v1             true         Bar
            bars                  v1             true         Bar
            bars                  v1             true         Bar
            bars                  v1             true         Bar
            bars                  v1             true         Bar
            bars                  v1             true         Bar
            foos     f,fo         v1             false        Foo
            foos     f,fo         v1             false        Foo
            foos     f,fo         v1             false        Foo
            foos     f,fo         v1             false        Foo
            foos     f,fo         v1             false        Foo
            foos     f,fo         v1             false        Foo
            foos     f,fo         v1             false        Foo
            foos     f,fo         v1             false        Foo
            foos     f,fo         v1             false        Foo
            bazzes   b            somegroup/v1   true         Baz
            bazzes   b            somegroup/v1   true         Baz
            bazzes   b            somegroup/v1   true         Baz
            bazzes   b            somegroup/v1   true         Baz
            bazzes   b            somegroup/v1   true         Baz
            bazzes   b            somegroup/v1   true         Baz
            bazzes   b            somegroup/v1   true         Baz
            bazzes   b            somegroup/v1   true         Baz
            bazzes   b            somegroup/v1   true         Baz
            bazzes   b            somegroup/v1   true         Baz
            bazzes   b            somegroup/v1   true         Baz
            bazzes   b            somegroup/v1   true         Baz
            bazzes   b            somegroup/v1   true         Baz
            bazzes   b            somegroup/v1   true         Baz

            expected: NAME     SHORTNAMES   APIVERSION     NAMESPACED   KIND
            bars                  v1             true         Bar
            foos     f,fo         v1             false        Foo
            bazzes   b            somegroup/v1   true         Baz
FAIL
FAIL	k8s.io/kubectl/pkg/cmd/apiresources	0.227s
FAIL
make: *** [Makefile:192: test] Error 1
```

</details>

brianpursley commented:
@dharmit I'll take a closer look at this PR later, but to answer your question about the failing unit test, it appears to be due to `resources` being a global variable.

Adding this at line 205 should prevent the error:
```
resources = []groupResource{}
```

brianpursley commented:
/ok-to-test

brianpursley commented on staging/src/k8s.io/kubectl/pkg/cmd/apiresources/apiresources_test.go:
```diff
@@ -211,6 +212,77 @@ bazzes.somegroup
+				cmd.Flags().Set("output", "json")
+			},
+			expectedOutput: `{
+	"kind": "v1",
```
Check your indentation on the `expectedOutput`.  I think it might need to be spaces instead of tabs in order to match the output, so this test case will pass.

dharmit replied:
That did the trick. Thanks 👍🏽 
What is the purpose of `expectedInvalidations` field there? I couldn't find anything.

brianpursley commented on staging/src/k8s.io/kubectl/pkg/cmd/apiresources/apiresources_printer.go:
```diff
@@ -0,0 +1,126 @@
+
+func (p APIResourcesPrinter) PrintObj(_ runtime.Object, writer io.Writer) error {
+	if p.noHeaders == false && p.outputFormat != "name" {
+		if err := printContextHeaders(writer, p.outputFormat); err != nil {
```
I think it would make sense to move `printContextHeaders` into this file, as this is the only place it is called.

dharmit replied:
Ack, makes sense. 👍🏽

brianpursley commented on staging/src/k8s.io/kubectl/pkg/cmd/apiresources/apiresources_test.go:
```diff
@@ -303,6 +375,7 @@ bazzes   b            somegroup/v1   true         Baz
 	}
 
 	for _, tc := range testCases {
+		resources = []groupResource{} // clean the package level variable to avoid stacking up of resources
```
I don't think this should be here.  The unit tests passed previously without it, so this would indicate a change in behavior.

There has to be another way to handle this without changing the test.

dharmit replied:
Agree. Moved it to where you suggested earlier.

brianpursley commented on staging/src/k8s.io/kubectl/pkg/cmd/apiresources/apiresources_printer.go:
```diff
@@ -0,0 +1,126 @@
+				name += "." + r.APIGroup
+			}
+			if _, err := fmt.Fprintf(writer, "%s\n", name); err != nil {
+				//errs = append(errs, err)
```
In the original code, errors were collected and an aggregate error was returned.

Can this code do the same thing as the original code, in order to preserve the existing behavior?

dharmit replied:
This is something I wanted an opinion on, so I left this part commented instead of removing it.

The only way I could think of preserving the existing behaviour was by introducing a goroutine and using an `error` channel to put each error on. Would that be reasonable?

brianpursley replied:
I don't think it should need channels.

It should be able to return an aggregate error, something like this:
```
        if len(errs) > 0 {
		return errors.NewAggregate(errs)
	}
```

brianpursley commented on staging/src/k8s.io/kubectl/pkg/cmd/apiresources/apiresources.go:
```diff
@@ -170,7 +192,15 @@ func (o *APIResourceOptions) RunAPIResources() error {
+		rl.APIResources = append(rl.APIResources, r.APIResources...)
+	}
 
 	for _, list := range lists {
```
Shouldn't this entire for loop be contained within `APIResourcesPrinter`?

It only applies to when you are not printing JSON or YAML, right?

Then I think that solves the `resources` global variable problem, because resources can remain a local variable inside `APIResourcesPrinter` now, if I am understanding this correctly.

dharmit replied:
This is actually the part that took me the longest while working on this issue. So I hope I understood things correctly.

So `lists` populated by the call
```go
lists, err := o.discoveryClient.ServerPreferredResources()
```
is a slice of type `metav1.APIResourceList`. Initially, I intended to use it directly in the respective `PrintObj` methods. But while `metav1.APIResourceList` itself implements the `runtime.Object` interface, a slice of it doesn't. Hence, it is how you see it right now.

I could possibly be missing something here. So please correct me if I am wrong. :)

brianpursley replied:
OK, I think I understand the reason.

You're re-purposing `metav1.APIResourceList` so that you have something that implements `runtime.Object` for the printer interface to work.

However, this doesn't seem like a good idea to me.  It is really a list of lists, so merging them all into a single APIResourceList is losing some necessary details. For example, I notice in the unit test for json, it is missing the group version (`v1` for Foo and Bar, and `somegroup/v1` for Baz and NoVerbs).

brianpursley commented on staging/src/k8s.io/kubectl/pkg/cmd/apiresources/apiresources.go:
```diff
@@ -208,44 +238,9 @@ func (o *APIResourceOptions) RunAPIResources() error {
-		}
-	}
-
 	sort.Stable(sortableResource{resources, o.SortBy})
```
Similar to my previous comment, I think this can be moved inside `APIResourcesPrinter` because it does not apply when you are printing JSON or YAML.

dharmit replied:
I think we need to keep it here, since there's no straightforward way of accessing the `o.SortBy` within `APIResourcePrinter` as it is right now. Of course, we could modify `APIResourcePrinter` to also include the `sortBy` field and then this could go there. I'll do as you suggest.

brianpursley replied:
Yeah, I think sorting would be an implementation detail of the printer, so it makes sense to have a `sortBy` field on the printer.

ardaguclu replied:
I think, there is no need sorting for json and yaml. We can add validation to not allow this and sorting can be moved under custom (default) printer.

brianpursley reviewed (changes requested):
Hi @dharmit I left a few comments for you to look into.

Ping me if you need clarification or if you make changes and want me to take another look.

dharmit commented:
> Hi @dharmit I left a few comments for you to look into.

@brianpursley thanks a lot for your feedback. 🙏🏽 

> Ping me if you need clarification or if you make changes and want me to take another look.

I have made a few changes and have a few questions for things I haven't changed. Your responses to how I could approach this would be greatly helpful.

brianpursley commented:
OK, it looks like the problem is that you've got `lists`, which is of type `[]*metav1.APIResourceList`, but like you said, it doesn't implement `runtime.Object` so it can't be passed to `PrintObj`.

I'm trying to think of a good way to get around this situation...

:thinking:

brianpursley commented:
I wonder if it even makes sense to try to use the `ResourcePrinter` interface for this command. And I say this because these records aren't actually resources, they are metadata about the known resource types, so I feel like this might be a case of trying to make something fit when it is not an exact fit.

Some other commands have alternate printers, like `kubectl top` and `kubectl explain`.

I wonder if this command could have its own printer that doesn't require a `runtime.Object` and uses `[]*metav1.APIResourceList` instead.

Check out [metrics_printer.go](https://github.com/kubernetes/kubernetes/blob/71acf56766c966653c3d1fe0ef6f3f036277d546/staging/src/k8s.io/kubectl/pkg/metricsutil/metrics_printer.go#L85) which `kubectl top` uses.  It's not exactly the same situation, but it is an example of what I'm talking out with a printer that does not implement `ResourcePrinter`.

dharmit commented:
> I wonder if it even makes sense to try to use the `ResourcePrinter` interface for this command. And I say this because these records aren't actually resources, they are metadata about the known resource types, so I feel like this might be a case of trying to make something fit when it is not an exact fit.

I get your point about trying to make something fit that's not an exact fit. What's also messed up in the current JSON output is that `apiVersion` shows as `APIResourceList` and `kind` shows `v1`. 🤦🏽‍♂️ 

> Some other commands have alternate printers, like `kubectl top` and `kubectl explain`.

Neither of these have JSON output. OTOH, `kubectl get` seems to have a repurposed printer too. Could that be a better reference? I have only skimmed through it while working on this one. Will check it in more detail now.

I'm literally brainstorming here so that I don't end up going down a rabbit hole, because I spent a few weeks before opening this PR. 😉

Also, @ardaguclu WDYT? Pinging you because you reviewed an earlier PR for this feature - https://github.com/kubernetes/kubernetes/pull/116926.

ardaguclu commented on staging/src/k8s.io/kubectl/pkg/cmd/apiresources/apiresources_printer.go:
```diff
@@ -0,0 +1,137 @@
+	noHeaders    bool
+}
+
+type PrintFlags struct {
```
No need to have `PrintFlags` in here, built-in PrintFlags already achieves this for us.

ardaguclu commented on staging/src/k8s.io/kubectl/pkg/cmd/apiresources/apiresources_printer.go:
```diff
@@ -0,0 +1,137 @@
+	OutputFormat       *string
+}
+
+func (f *PrintFlags) AllowedFormats() []string {
```
No need to have `AllowedFormats` in here and probably the reason of the usage of JSONYamlPrintFlags. We can manually check this in the code.

ardaguclu commented on staging/src/k8s.io/kubectl/pkg/cmd/apiresources/apiresources_printer.go:
```diff
@@ -0,0 +1,137 @@
+	return p, nil
+}
+
+func (f *PrintFlags) AddFlags(cmd *cobra.Command) {
```
same, no need

ardaguclu commented on staging/src/k8s.io/kubectl/pkg/cmd/apiresources/apiresources_test.go:
```diff
@@ -208,6 +209,78 @@ bazzes   b            somegroup/v1   true         Baz    get,list,create,delete
+`,
+			expectedInvalidations: 1,
+		},
+		{
```
Thank you for this unit test. But we definitely need more.

ardaguclu commented on staging/src/k8s.io/kubectl/pkg/cmd/apiresources/apiresources.go:
```diff
@@ -170,8 +192,17 @@ func (o *APIResourceOptions) RunAPIResources() error {
 	}
 
-	resources := []groupResource{}
+	rl := &metav1.APIResourceList{
```
`lists` returned from discovery should satisfy the printobject's requirements?. I mean what about something like this:

```go
	resources := []groupResource{}
	var filteredList []*metav1.APIResourceList

	for _, list := range lists {
		if len(list.APIResources) == 0 {
			continue
		}
		gv, err := schema.ParseGroupVersion(list.GroupVersion)
		if err != nil {
			continue
		}
		apilist := &metav1.APIResourceList{
			TypeMeta: metav1.TypeMeta{
				Kind:       "v1",
				APIVersion: "APIResourceList",
			},
			GroupVersion: gv.String(),
		}
		var apiresources []metav1.APIResource
		for _, resource := range list.APIResources {
			if len(resource.Verbs) == 0 {
				continue
			}
			// filter apiGroup
			if o.groupChanged && o.APIGroup != gv.Group {
				continue
			}
			// filter namespaced
			if o.nsChanged && o.Namespaced != resource.Namespaced {
				continue
			}
			// filter to resources that support the specified verbs
			if len(o.Verbs) > 0 && !sets.NewString(resource.Verbs...).HasAll(o.Verbs...) {
				continue
			}
			// filter to resources that belong to the specified categories
			if len(o.Categories) > 0 && !sets.NewString(resource.Categories...).HasAll(o.Categories...) {
				continue
			}

			resources = append(resources, groupResource{
				APIGroup:        gv.Group,
				APIGroupVersion: gv.String(),
				APIResource:     resource,
			})
			apiresources = append(apiresources, resource)
		}
		if len(apiresources) > 0 {
			apilist.APIResources = apiresources
			filteredList = append(filteredList, apilist)
		}
	}

	// if output format is json or yaml, directly print filteredList and short cut

	// if output format is default, print resources as always

	if o.NoHeaders == false && o.Output != "name" {
```

dharmit replied:
> `lists` returned from discovery should satisfy the printobject's requirements?

I don't think so. And that's why I went with the complicated looking approach. `lists` is a slice of type `APIResourceList`. While `APIResourceList` itself implements the `runtime.Object` interface, its slice `[]*metav1.APIResourceList` doesn't.

dharmit replied:
@ardaguclu I have implemented it the way you suggested in this comment. Look at 4e52bc3. IMO, the problem with this approach is that it prints a bunch of JSON object not a unified JSON object.

So in the next commit 562d48a I flattened the `filteredList` and print a unified JSON object. However, it doesn't have the `groupVersion` info that we presently print under the `APIVersion` column.

This is the reason I spent too long before opening the PR. In hindsight, I feel that I should have brought it up earlier.

Note that there are no tests yet. I would like to get feedback on what we want the output to be like, and then I'll focus on adding tests.

cc @brianpursley

dharmit committed:
JSON & YAML output for kubectl api-resources

Signed-off-by: Dharmit Shah <shahdharmit@gmail.com>

ardaguclu commented on staging/src/k8s.io/kubectl/pkg/cmd/apiresources/apiresources.go:
```diff
@@ -151,6 +154,18 @@ func (o *APIResourceOptions) Complete(restClientGetter genericclioptions.RESTCli
 
+	var printer printers.ResourcePrinter
+	if o.PrintFlags.OutputFormat != nil && len(*o.PrintFlags.OutputFormat) > 0 {
+		if *o.PrintFlags.OutputFormat == "json" || *o.PrintFlags.OutputFormat == "yaml" {
```
nit: It is better to merge this with the above in one if statement.

dharmit replied:
Ack, done.

ardaguclu commented on staging/src/k8s.io/kubectl/pkg/cmd/apiresources/apiresources.go:
```diff
@@ -205,18 +229,36 @@ func (o *APIResourceOptions) RunAPIResources() error {
 
-	if o.NoHeaders == false && o.Output != "name" {
-		if err = printContextHeaders(w, o.Output); err != nil {
+	if o.NoHeaders == false && *o.PrintFlags.OutputFormat != "name" && (*o.PrintFlags.OutputFormat != "json" && *o.PrintFlags.OutputFormat != "yaml") {
```
Why don't we do only `*o.PrintFlags.OutputFormat == ""`

dharmit replied:
Modified this, but it needed more checks than that. PTAL.

ardaguclu reviewed (commented):
This mechanism looks good to me. We definitely need unit tests and integration test to verify the behavior. Dropped a few comments as well. Thank you.

dharmit committed:
Unit tests for JSON & YAML outputs

Signed-off-by: Dharmit Shah <shahdharmit@gmail.com>

dharmit commented:
> This mechanism looks good to me. We definitely need unit tests and integration test to verify the behavior.

@ardaguclu I have added unit tests for JSON & YAML output and also one more that I thought was relevant to cover. PTAL. Please also suggest me if more unit test needs to be added. I couldn't think of more.

As for integration tests, can you point me where I should look for example tests for `kubectl api-resources`? I couldn't find any in `test/e2e/kubectl/`.

dharmit commented:
@ardaguclu I have updated the PR based on your feedback and added unit tests. Just want to point out that @brianpursley mentioned in the past that the JSON/YAML outputs should preferably have that.

However, we're creating that groupVersion info [here](https://github.com/kubernetes/kubernetes/blob/85ff7e75a2066fef3543a55ca995c5173e779b72/staging/src/k8s.io/kubectl/pkg/cmd/apiresources/apiresources.go#L179). I don't see how we could have it in the current approach of flattening the `[]metav1.APIResourceList`.

Otherwise, I think, this PR is ready for review.

dharmit commented:
Regarding the failing linter hints. The first error is:
```
ERROR: staging/src/k8s.io/kubectl/pkg/cmd/apiresources/apiresources.go:239:5: S1002: should omit comparison to bool constant, can be simplified to `!o.NoHeaders` (gosimple)
ERROR: 	if o.NoHeaders == false && *o.PrintFlags.OutputFormat == "" || (*o.PrintFlags.OutputFormat == "wide" && !o.NoHeaders) { 
```
We're checking `o.NoHeaders == false` already even outside this PR. Wonder why it's failing now. 🤔 

And the other three errors are with unit tests:
```
ERROR: staging/src/k8s.io/kubectl/pkg/cmd/apiresources/apiresources_test.go:206:20: Error return value of `(*github.com/spf13/pflag.FlagSet).Set` is not checked (errcheck)
ERROR: 				cmd.Flags().Set("output", "wide")
ERROR: 				               ^
ERROR: staging/src/k8s.io/kubectl/pkg/cmd/apiresources/apiresources_test.go:207:20: Error return value of `(*github.com/spf13/pflag.FlagSet).Set` is not checked (errcheck)
ERROR: 				cmd.Flags().Set("no-headers", "true")
ERROR: 				               ^
ERROR: staging/src/k8s.io/kubectl/pkg/cmd/apiresources/apiresources_test.go:318:20: Error return value of `(*github.com/spf13/pflag.FlagSet).Set` is not checked (errcheck)
ERROR: 				cmd.Flags().Set("output", "json")
ERROR: 				               ^
ERROR: staging/src/k8s.io/kubectl/pkg/cmd/apiresources/apiresources_test.go:380:20: Error return value of `(*github.com/spf13/pflag.FlagSet).Set` is not checked (errcheck)
ERROR: 				cmd.Flags().Set("output", "yaml")
ERROR: 				               ^
```
But the error return value of other test cases is also not being checked, AFAICS.

What am I missing here? I'll change the first one from `o.NoHeaders == false` to `!o.NoHeaders`, but what do I need to do about the issue it's pointing out in the unit tests?

Also, what is the meaning of `expectedInvalidations` in the test struct?

MikeSpreitzer commented on staging/src/k8s.io/kubectl/pkg/cmd/apiresources/apiresources.go:
```diff
@@ -151,6 +154,17 @@ func (o *APIResourceOptions) Complete(restClientGetter genericclioptions.RESTCli
+		if err != nil {
+			return err
+		}
+		o.PrintObj = func(object runtime.Object, out io.Writer) error {
+			return printer.PrintObj(object, out)
+		}
```
```suggestion
		o.PrintObj = printer.PrintObj
```

MikeSpreitzer commented on staging/src/k8s.io/kubectl/pkg/cmd/apiresources/apiresources.go:
```diff
@@ -205,18 +228,36 @@ func (o *APIResourceOptions) RunAPIResources() error {
 
-	if o.NoHeaders == false && o.Output != "name" {
-		if err = printContextHeaders(w, o.Output); err != nil {
+	if o.NoHeaders == false && *o.PrintFlags.OutputFormat == "" || (*o.PrintFlags.OutputFormat == "wide" && !o.NoHeaders) {
```
```suggestion
	if o.NoHeaders == false && (*o.PrintFlags.OutputFormat == "" || *o.PrintFlags.OutputFormat == "wide") {
```

dharmit replied:
Thanks. My local branch had the `!o.NoHeaders` part of it already, but not the correct OR check that you provided. Done.

MikeSpreitzer commented:
In the unit tests I do not see API group and version. Those are part of the existing output formats, and should be present in the JSON and YAML formats too.

dharmit commented:
> In the unit tests I do not see API group and version. Those are part of the existing output formats, and should be present in the JSON and YAML formats too.

To generate the JSON and YAML outputs, we are flattening the slice of `APIResourceList` we get here: https://github.com/kubernetes/kubernetes/blob/6cc357046615410434b2fa99c5a229e5db2eb6d9/staging/src/k8s.io/kubectl/pkg/cmd/apiresources/apiresources.go#L168

We are flattening it because each element of that slice contains a slice of `APIResource`.

Now, the `GroupVersion` information is available per `APIResourceList` whereas the JSON and YAML outputs in the PR's current form are giving us `APIResource`. And the reason we are doing this is that we want to reuse existing JSONYAMLPrinter instead of implementing a custom struct. Ref: https://github.com/kubernetes/kubernetes/pull/116926#pullrequestreview-1357999561. A `[]APIResourceList` doesn't implement the required interface, whereas `APIResourceList` does.

From a user perspective, I understand that this might not be ideal output. But I can't think of a way here without creating a new struct that also implements the `ResourcePrinter` interface, which requires implementing the `runtime.Object` interface: https://github.com/kubernetes/kubernetes/blob/6cc357046615410434b2fa99c5a229e5db2eb6d9/staging/src/k8s.io/cli-runtime/pkg/printers/interface.go#L34-L38

PS: I don't expect to have done a good job at explaining the dilemma here, but I'm willing to make changes that seem necessary here. @ardaguclu could you please chime in? :)

soltysh commented on staging/src/k8s.io/kubectl/pkg/cmd/apiresources/apiresources.go:
```diff
@@ -123,9 +126,9 @@ func NewCmdAPIResources(restClientGetter genericclioptions.RESTClientGetter, ioS
-	supportedOutputTypes := sets.New[string]("", "wide", "name")
-	if !supportedOutputTypes.Has(o.Output) {
-		return fmt.Errorf("--output %v is not available", o.Output)
+	supportedOutputTypes := sets.NewString("", "wide", "name", "json", "yaml")
```
You've decided to use entire [PrintFlags](https://github.com/kubernetes/kubernetes/blob/e319c541f144e9bee6160f1dd8671638a9029f4c/staging/src/k8s.io/cli-runtime/pkg/genericclioptions/print_flags.go#L63-L65) which includes json/yaml printer, name printer and also template printer. Each one of them adds its own flags, but also has the ability to validate the contents of `--output` flag. With this check you're basically eliminating that [last printer](https://github.com/kubernetes/kubernetes/blob/e319c541f144e9bee6160f1dd8671638a9029f4c/staging/src/k8s.io/cli-runtime/pkg/genericclioptions/kube_template_flags.go#L28) which I'm not sure we want here in the first place. 

So my question for you is, which approach you prefer to have all of the printers (including the templating ones as well) or only the json/yaml and name printers? In either case this check is not necessary, since when creating the printer you'll get appropriate error if `--output` flag is not one of the accepted ones.

dharmit replied:
> So my question for you is, which approach you prefer to have all of the printers (including the templating ones as well) or only the json/yaml and name printers?

Based on what you mentioned, I'd go with only the JSON/YAML printers and skip the rest.

> In either case this check is not necessary, since when creating the printer you'll get appropriate error if --output flag is not one of the accepted ones.

But there already was a check for `sets.New[string]("", "wide", "name")`. Don't we need these? If no, how do we validate this?

Besides, we don't have `OutputFormat` field any more if we go with `JSONYamlPrintFlags`. What I can't wrap my head around is how do we make the user choose from these options:
- wide
- name
- json
- yaml

The reason I can't wrap my head around it is that while user can do `-o wide` and `-o name`, these are not delegated to any of the printers in `staging/src/k8s.io/cli-runtime/pkg/printers`, but implemented using `switch..case`: https://github.com/kubernetes/kubernetes/blob/fd15e3fd5566fb0a65ded1883fbf51ce7a68fe28/staging/src/k8s.io/kubectl/pkg/cmd/apiresources/apiresources.go#L219-L239

soltysh commented on staging/src/k8s.io/kubectl/pkg/cmd/apiresources/apiresources.go:
```diff
@@ -151,6 +154,17 @@ func (o *APIResourceOptions) Complete(restClientGetter genericclioptions.RESTCli
+	var printer printers.ResourcePrinter
+	if o.PrintFlags.OutputFormat != nil && len(*o.PrintFlags.OutputFormat) > 0 &&
+		(*o.PrintFlags.OutputFormat == "json" || *o.PrintFlags.OutputFormat == "yaml") {
+		printer, err = o.PrintFlags.ToPrinter()
```
Based on what I wrote above it's sufficient to squash this condition to just:
```golang
if o.PrintFlags.OutputFormat != nil {
    printer, err = o.PrintFlags.ToPrinter()
    ...
```
since the rest will be handled by the printers, including the validation I wrote above.

soltysh commented on staging/src/k8s.io/kubectl/pkg/cmd/apiresources/apiresources.go:
```diff
@@ -205,18 +228,36 @@ func (o *APIResourceOptions) RunAPIResources() error {
 			})
+			apiResources = append(apiResources, resource)
+		}
+		if len(apiResources) > 0 {
```
Don't filter out empty groups, this will cause that you won't print groups that don't have any resources. Filtering is happening earlier (there are `continue` blocks ensuring you'll skip group you don't care about).

soltysh commented on staging/src/k8s.io/kubectl/pkg/cmd/apiresources/apiresources.go:
```diff
@@ -171,6 +185,7 @@ func (o *APIResourceOptions) RunAPIResources() error {
 	}
 
 	resources := []groupResource{}
+	var filteredList []*metav1.APIResourceList
```
Can you rename the `resources` to `allResources`, so it's clear which one is what.

soltysh commented on staging/src/k8s.io/kubectl/pkg/cmd/apiresources/apiresources.go:
```diff
@@ -205,18 +228,36 @@ func (o *APIResourceOptions) RunAPIResources() error {
 
-	if o.NoHeaders == false && o.Output != "name" {
-		if err = printContextHeaders(w, o.Output); err != nil {
+	if !o.NoHeaders && (*o.PrintFlags.OutputFormat == "" || *o.PrintFlags.OutputFormat == "wide") {
```
```suggestion
	if !o.NoHeaders && (o.PrintFlags.OutputFormat == nil || *o.PrintFlags.OutputFormat == "" || *o.PrintFlags.OutputFormat == "wide") {
```

soltysh commented on staging/src/k8s.io/kubectl/pkg/cmd/apiresources/apiresources.go:
```diff
@@ -205,18 +228,36 @@ func (o *APIResourceOptions) RunAPIResources() error {
 	}
 
+	if *o.PrintFlags.OutputFormat == "json" || *o.PrintFlags.OutputFormat == "yaml" {
+		flatList := &metav1.APIResourceList{
```
What is the reason you can't create this `flatList` already when creating `filteredList`? With appropriate comment above that variable.

soltysh reviewed (changes requested):
After reviewing the entire PR I'm more convinced that rather than using entire `PrintFlags` struct, you should only add json/yaml printer, since you're not using anything else other than just those two printers. Even more, rather than wiring the flags structure, I'd advise you to just inject the yaml/json printer directly based on the `--output` flag value provided by the user. I think this should simplify your implementation a bit.

dharmit commented:
/remove-lifecycle stale

Will get onto this soon. Still very much want to get this merged.

dharmit commented:
@soltysh besides asking this question (https://github.com/kubernetes/kubernetes/pull/126338#discussion_r2086442802) I have pushed a [separate branch](https://github.com/dharmit/kubernetes/tree/1382-printers) which contains this commit https://github.com/dharmit/kubernetes/commit/52f7ccf9758f002d9c1bb944fdfe4f5412a4e400. It uses a separate `PrintFlags` within `apiresources.go` similar to how `kubectl get` does it. I feel it's a cleaner way of implementing the asked feature. I can add tests to it if you think it makes more sense than that PR.

Let me know what you think.

cc @ardaguclu @brianpursley @MikeSpreitzer

soltysh commented:
> @soltysh besides asking this question ([#126338 (comment)](https://github.com/kubernetes/kubernetes/pull/126338#discussion_r2086442802)) I have pushed a [separate branch](https://github.com/dharmit/kubernetes/tree/1382-printers) which contains this commit [dharmit@52f7ccf](https://github.com/dharmit/kubernetes/commit/52f7ccf9758f002d9c1bb944fdfe4f5412a4e400). It uses a separate `PrintFlags` within `apiresources.go` similar to how `kubectl get` does it. I feel it's a cleaner way of implementing the asked feature. I can add tests to it if you think it makes more sense than that PR.
> 
> Let me know what you think.

I really like your approach! 

I think the only suggestion I have is to add something similar to [HumanReadableFlags](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/kubectl/pkg/cmd/get/humanreadable_flags.go) to handle both the name and wide options. Just to make it more composable.

dharmit commented:
@soltysh Adding the HumanReadableFlags and NamePrintFlags to the mix is causing an issue with the `flatList` in that both the printers are exiting with the error indicating `metav1.Object` interface is not implemented. And I'm feeling lost as to how to address this.

Can you help me figure what I'm doing wrong or what I could try? I have pushed the code as a separate commit https://github.com/dharmit/kubernetes/commit/42321822dcc814e5a6c3469e5302384aa25c94f7.

soltysh commented:
> Can you help me figure what I'm doing wrong or what I could try? I have pushed the code as a separate commit [dharmit@4232182](https://github.com/dharmit/kubernetes/commit/42321822dcc814e5a6c3469e5302384aa25c94f7).

As mentioned on slack: 
1. don't re-use `HumanReadableFlags`, I've meant more to copy the pattern, once we have the shape, we can consider moving that printer to shared printers, but I doubt it, since the one from get will differ from the one in apiresources;
2. very likely similar problem is with `NamePrintFlags`, which is not quite usable in apiresources; 

I haven't looked too closely into what kind of error you're getting, but I'm suspecting this is coming from that. Ideally I'd suggest either opening entirely new PR or update the existing one with that code and let's continue discussion there.

dharmit commented:
Closing in favour of https://github.com/kubernetes/kubernetes/pull/132604.
